### PR TITLE
RV-ify balding nichols

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
This kinda depends on #2480

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 1]
```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
     val Fst1_k = (1d - Fst_k) :/ Fst_k
     val Fst1_kBc = sc.broadcast(Fst1_k)
 
-    val rdd = sc.parallelize((0 until M).map(m => (m, Rand.randInt.draw())), nPartitions)
-      .map { case (m, perVariantSeed) =>
-        val perVariantRandomGenerator = new JDKRandomGenerator
-        perVariantRandomGenerator.setSeed(perVariantSeed)
-        val perVariantRandomBasis = new RandBasis(perVariantRandomGenerator)
+    val rowType = TStruct(
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: Don't duplicate this, hoist up the matrix type and use `matrixType.rowType`.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 2]
```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
...
+      "pk" -> TLocus(gr),
+      "v" -> TVariant(gr),
+      "va" -> TStruct("ancestralAF" -> TFloat64(), "AF" -> TArray(TFloat64())),
+      "gs" -> !TArray(TStruct("GT" -> TCall())))
 
-        val ancestralAF = af_dist.getBreezeDist(perVariantRandomBasis).draw()
 
-        val popAF_k = (0 until K).map { k =>
-          new Beta(ancestralAF * Fst1_kBc.value(k), (1 - ancestralAF) * Fst1_kBc.value(k))(perVariantRandomBasis).draw()
+    val rdd = sc.parallelize((0 until M).map(m => (m, Rand.randInt.draw())), nPartitions)
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: This creates a Vector.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 3]
```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
...
+                1
+
+            rvb.startStruct()
+            rvb.addInt(gt)
+            rvb.endStruct()
+            i += 1
+          }
+          rvb.endArray()
+          rvb.endStruct()
+          rvb.result()
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: We should delete this.  It conses.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 4]
```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
...
+        val rvb = new RegionValueBuilder(MemoryBuffer())
+
+        it.map { case (m, perVariantSeed) =>
+          val perVariantRandomGenerator = new JDKRandomGenerator
+          perVariantRandomGenerator.setSeed(perVariantSeed)
+          val perVariantRandomBasis = new RandBasis(perVariantRandomGenerator)
+
+          val ancestralAF = af_dist.getBreezeDist(perVariantRandomBasis).draw()
+
+          val popAF_k = (0 until K).map { k =>
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: Vector.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 5]
```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
...
+          rvb.endArray()
+          rvb.endStruct()
+
+          // gs
+          rvb.startArray(N)
+          i = 0
+          while (i < N) {
+            val p = popAF_k(popOfSample_nBc.value(i))
+            val pSq = p * p
+            val x = new Uniform(0, 1)(perVariantRandomBasis).draw()
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: You can joist this out and just call `draw`. ehre.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala (hunk 3 above)
Comment (tpoterba) [reply src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: What's the preferred alternative? We could also allocate a region value once in the builder and mutate it.

Comment (cseed) [reply src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: Yeah, allocate a region value in the mapPartitions, and then:

```
  rv.setOffset(rvb.end())
  rv
```

There's also `RegionValue.set(region: MemoryBuffer, offset: Long)`.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 6]
```diff
@@ -63,7 +64,7 @@ object BaldingNicholsModel {
     val popDist = DenseVector(popDistArray)
     val FstOfPop = DenseVector(FstOfPopArray)
 
-    info(s"baldingnichols: generating genotypes for $K populations, $N samples, and $M variants...")
+    info(s"balding nichols: generating genotypes for $K populations, $N samples, and $M variants...")
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: This should match the python name: `balding_nichols_model`.

File: src/main/scala/is/hail/stats/BaldingNicholsModel.scala [hunk 7]
```diff
@@ -135,8 +105,110 @@ object BaldingNicholsModel {
...
+      vaSignature,
+      globalSignature,
+      TStruct("GT" -> TCall()),
+      wasSplit = true)
+
+    val matrixType = MatrixType(vsmMetadata)
+
+    val rowType = matrixType.rowType
+
+
```
Comment (cseed) [inline src/main/scala/is/hail/stats/BaldingNicholsModel.scala]: Extra space.

**Pull request merged**